File: projects/project-02-agent-readable-workspace/starter/python/main.py
"""kb v1, project 02 Python starter.

This is project 01's solution app (init, list, ask, serve), the state a
new session inherits. Project 02's work is the delta in SPEC.md: import,
the show detail view, the metadata index as system of record, the
document-detail endpoint, and workspace-check. Run ./verify.sh until the
solution-stage cases pass against your changes.
"""

from __future__ import annotations

import json
import re
import shutil
import sys
import threading
import urllib.request
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer
from pathlib import Path
from urllib.parse import parse_qs, urlparse

MIN_TOKEN_LENGTH = 4
MAX_CITATIONS = 2
NO_MATCH_ANSWER = (
    "No matching lines in the document set. "
    "Import more documents or rephrase the question."
)
USAGE = (
    "usage: kb init --data-dir DIR [--seed SRC] | list --data-dir DIR | "
    'ask --data-dir DIR "QUESTION" | serve --data-dir DIR [--port N] [--self-check]'
)


def uninitialized_error(data_dir: str) -> str:
    return f"error: data directory {data_dir} is not initialized; run kb init first"


# ---------------------------------------------------------------- documents


def tokenize(text: str) -> list[str]:
    return re.findall(rf"[a-z0-9]{{{MIN_TOKEN_LENGTH},}}", text.lower())


def load_documents(data_dir: Path) -> list[dict]:
    documents = []
    for path in sorted((data_dir / "documents").iterdir()):
        if path.suffix not in (".md", ".txt") or not path.is_file():
            continue
        text = path.read_text(encoding="utf-8")
        lines = text.rstrip("\n").split("\n")
        title = path.name
        for line in lines:
            if line.startswith("# "):
                title = line[2:].strip()
                break
        documents.append(
            {"id": path.stem, "title": title, "filename": path.name, "lines": lines}
        )
    return documents


def is_initialized(data_dir: Path) -> bool:
    return (data_dir / "documents").is_dir()


# ----------------------------------------------------------------- commands


def cmd_init(data_dir_arg: str, seed: str | None) -> tuple[int, str, str]:
    data_dir = Path(data_dir_arg)
    created = []
    for directory in (data_dir, data_dir / "documents", data_dir / "index"):
        if not directory.is_dir():
            directory.mkdir(parents=True)
            created.append(directory.as_posix())
    seeded = []
    if seed is not None:
        seed_dir = Path(seed)
        if not seed_dir.is_dir():
            return 2, "", f"error: cannot read seed directory {seed}"
        for source in sorted(seed_dir.iterdir()):
            if source.suffix not in (".md", ".txt") or not source.is_file():
                continue
            target = data_dir / "documents" / source.name
            if not target.is_file():
                shutil.copyfile(source, target)
                seeded.append(source.name)
    report = {"data_dir": data_dir.as_posix(), "created": created, "seeded": seeded}
    return 0, json.dumps(report, indent=2) + "\n", ""


def cmd_list(data_dir_arg: str) -> tuple[int, str, str]:
    data_dir = Path(data_dir_arg)
    if not is_initialized(data_dir):
        return 1, "", uninitialized_error(data_dir_arg)
    documents = [
        {
            "id": doc["id"],
            "title": doc["title"],
            "filename": doc["filename"],
            "lines": len(doc["lines"]),
        }
        for doc in load_documents(data_dir)
    ]
    return 0, json.dumps({"documents": documents}, indent=2) + "\n", ""


def retrieve(documents: list[dict], question: str) -> list[dict]:
    question_tokens = set(tokenize(question))
    candidates = []
    for doc in documents:
        for number, line in enumerate(doc["lines"], start=1):
            stripped = line.strip()
            if not stripped:
                continue
            score = len(question_tokens & set(tokenize(stripped)))
            if score > 0:
                candidates.append(
                    {
                        "document": doc["id"],
                        "title": doc["title"],
                        "line": number,
                        "excerpt": stripped,
                        "score": score,
                    }
                )
    candidates.sort(key=lambda c: (-c["score"], c["document"], c["line"]))
    return candidates[:MAX_CITATIONS]


def compose_answer(citations: list[dict]) -> str:
    """The model seam: a real assistant would generate prose here. This
    deterministic composer quotes the best citation instead, keeping the
    citation contract identical for both."""
    if not citations:
        return NO_MATCH_ANSWER
    first = citations[0]
    answer = f'Based on "{first["title"]}" (line {first["line"]}): {first["excerpt"]}'
    if len(citations) > 1:
        second = citations[1]
        answer += f' See also "{second["title"]}" (line {second["line"]}).'
    return answer


def cmd_ask(data_dir_arg: str, question: str) -> tuple[int, str, str]:
    data_dir = Path(data_dir_arg)
    if not is_initialized(data_dir):
        return 1, "", uninitialized_error(data_dir_arg)
    citations = retrieve(load_documents(data_dir), question)
    report = {
        "question": question,
        "citations": citations,
        "answer": compose_answer(citations),
    }
    return 0, json.dumps(report, indent=2) + "\n", ""


# -------------------------------------------------------------------- serve


def health_payload(data_dir: Path) -> dict:
    return {"status": "ok", "documents": len(load_documents(data_dir))}


class KbHandler(BaseHTTPRequestHandler):
    data_dir: Path  # assigned by make_server

    def _send_json(self, status: int, payload: dict) -> None:
        body = json.dumps(payload, indent=2).encode("utf-8")
        self.send_response(status)
        self.send_header("Content-Type", "application/json")
        self.send_header("Content-Length", str(len(body)))
        self.end_headers()
        self.wfile.write(body)

    def do_GET(self) -> None:  # noqa: N802 (http.server API name)
        parsed = urlparse(self.path)
        if parsed.path == "/health":
            self._send_json(200, health_payload(self.data_dir))
        elif parsed.path == "/documents":
            _, out, _ = cmd_list(self.data_dir.as_posix())
            self._send_json(200, json.loads(out))
        elif parsed.path == "/ask":
            question = parse_qs(parsed.query).get("q", [""])[0]
            _, out, _ = cmd_ask(self.data_dir.as_posix(), question)
            self._send_json(200, json.loads(out))
        else:
            self._send_json(404, {"error": "not found"})

    def log_message(self, format: str, *args: object) -> None:
        sys.stderr.write("kb serve: " + format % args + "\n")


def cmd_serve(data_dir_arg: str, port: int, self_check: bool) -> tuple[int, str, str]:
    data_dir = Path(data_dir_arg)
    if not is_initialized(data_dir):
        return 1, "", uninitialized_error(data_dir_arg)
    handler = type("BoundHandler", (KbHandler,), {"data_dir": data_dir})
    server = ThreadingHTTPServer(("127.0.0.1", port), handler)
    if not self_check:
        address = f"http://127.0.0.1:{server.server_address[1]}"
        print(f"kb serve: listening on {address}", file=sys.stderr)
        try:
            server.serve_forever()
        except KeyboardInterrupt:
            pass
        finally:
            server.server_close()
        return 0, "", ""
    thread = threading.Thread(target=server.serve_forever, daemon=True)
    thread.start()
    try:
        base = f"http://127.0.0.1:{server.server_address[1]}"
        with urllib.request.urlopen(base + "/health") as response:
            health = json.loads(response.read().decode("utf-8"))
        with urllib.request.urlopen(base + "/documents") as response:
            documents = json.loads(response.read().decode("utf-8"))
    finally:
        server.shutdown()
        server.server_close()
    report = {"self_check": {"health": health, "documents": len(documents["documents"])}}
    return 0, json.dumps(report, indent=2) + "\n", ""


# ---------------------------------------------------------------------- cli


def parse_flags(argv: list[str]) -> tuple[dict, list[str]]:
    flags = {}
    positional = []
    index = 0
    while index < len(argv):
        arg = argv[index]
        if arg in ("--data-dir", "--seed", "--port"):
            if index + 1 >= len(argv):
                raise ValueError(f"missing value for {arg}")
            flags[arg] = argv[index + 1]
            index += 2
        elif arg == "--self-check":
            flags[arg] = "true"
            index += 1
        else:
            positional.append(arg)
            index += 1
    return flags, positional


def dispatch(argv: list[str]) -> tuple[int, str, str]:
    if not argv:
        return 2, "", USAGE
    command, rest = argv[0], argv[1:]
    try:
        flags, positional = parse_flags(rest)
    except ValueError as error:
        return 2, "", f"error: {error}"
    if command == "init" and "--data-dir" in flags and not positional:
        return cmd_init(flags["--data-dir"], flags.get("--seed"))
    if command == "list" and "--data-dir" in flags and not positional:
        return cmd_list(flags["--data-dir"])
    if command == "ask" and "--data-dir" in flags and len(positional) == 1:
        return cmd_ask(flags["--data-dir"], positional[0])
    if command == "serve" and "--data-dir" in flags and not positional:
        port_text = flags.get("--port", "0")
        if not port_text.isdigit():
            return 2, "", f"error: invalid port {port_text}"
        return cmd_serve(flags["--data-dir"], int(port_text), "--self-check" in flags)
    return 2, "", USAGE


def main(argv: list[str]) -> int:
    exit_code, out, err = dispatch(argv[1:])
    if out:
        sys.stdout.write(out)
    if err:
        print(err, file=sys.stderr)
    return exit_code


if __name__ == "__main__":
    sys.exit(main(sys.argv))

File: projects/project-02-agent-readable-workspace/starter/typescript/main.ts
// kb v1, project 02 TypeScript starter.
//
// This is project 01's solution app (init, list, ask, serve), the state a
// new session inherits. Project 02's work is the delta in SPEC.md: import,
// the show detail view, the metadata index as system of record, the
// document-detail endpoint, and workspace-check. Run ./verify.sh until the
// solution-stage cases pass against your changes.

import {
  copyFileSync,
  existsSync,
  mkdirSync,
  readdirSync,
  readFileSync,
  statSync,
} from "node:fs";
import { createServer, type Server } from "node:http";
import { join } from "node:path";
import { pathToFileURL } from "node:url";

// Python's read_text() applies universal newlines: CRLF and CR become LF
// before anything else sees the text. readFileSync does not, so a document
// authored on Windows split into different lines, paragraphs and chunks in
// this track than in the Python one, and even hashed differently, since
// sha256Text hashes the text rather than the bytes. Every text read goes
// through here so both tracks see identical input. Byte-level reads (sha
// digests over file bytes, the guard's containment probe) deliberately do
// not, and stay on readFileSync.
function readText(path: string): string {
  return readFileSync(path, "utf8").replace(/\r\n?/g, "\n");
}


const MIN_TOKEN_LENGTH = 4;
const MAX_CITATIONS = 2;
const NO_MATCH_ANSWER =
  "No matching lines in the document set. Import more documents or rephrase the question.";
const USAGE =
  "usage: kb init --data-dir DIR [--seed SRC] | list --data-dir DIR | " +
  'ask --data-dir DIR "QUESTION" | serve --data-dir DIR [--port N] [--self-check] | ' +
  "experiment [--workdir DIR]";

type CommandResult = [exitCode: number, stdout: string, stderr: string];

function uninitializedError(dataDir: string): string {
  return `error: data directory ${dataDir} is not initialized; run kb init first`;
}

// ---------------------------------------------------------------- documents

export interface LoadedDocument {
  id: string;
  title: string;
  filename: string;
  lines: string[];
}

export function tokenize(text: string): string[] {
  const pattern = new RegExp(`[a-z0-9]{${MIN_TOKEN_LENGTH},}`, "g");
  return text.toLowerCase().match(pattern) ?? [];
}

function loadDocuments(dataDir: string): LoadedDocument[] {
  const documentsDir = join(dataDir, "documents");
  const documents: LoadedDocument[] = [];
  for (const name of readdirSync(documentsDir).sort()) {
    if (!name.endsWith(".md") && !name.endsWith(".txt")) {
      continue;
    }
    const path = join(documentsDir, name);
    if (!statSync(path).isFile()) {
      continue;
    }
    const text = readText(path);
    const lines = text.replace(/\n+$/, "").split("\n");
    let title = name;
    for (const line of lines) {
      if (line.startsWith("# ")) {
        title = line.slice(2).trim();
        break;
      }
    }
    documents.push({ id: name.replace(/\.[^.]+$/, ""), title, filename: name, lines });
  }
  return documents;
}

function isInitialized(dataDir: string): boolean {
  return existsSync(join(dataDir, "documents")) && statSync(join(dataDir, "documents")).isDirectory();
}

// ----------------------------------------------------------------- commands

export function cmdInit(dataDir: string, seed: string | null): CommandResult {
  const created: string[] = [];
  for (const directory of [dataDir, join(dataDir, "documents"), join(dataDir, "index")]) {
    if (!existsSync(directory)) {
      mkdirSync(directory, { recursive: true });
      created.push(directory.split("\\").join("/"));
    }
  }
  const seeded: string[] = [];
  if (seed !== null) {
    if (!existsSync(seed) || !statSync(seed).isDirectory()) {
      return [2, "", `error: cannot read seed directory ${seed}`];
    }
    for (const name of readdirSync(seed).sort()) {
      if (!name.endsWith(".md") && !name.endsWith(".txt")) {
        continue;
      }
      const source = join(seed, name);
      const target = join(dataDir, "documents", name);
      if (statSync(source).isFile() && !existsSync(target)) {
        copyFileSync(source, target);
        seeded.push(name);
      }
    }
  }
  const report = { data_dir: dataDir.split("\\").join("/"), created, seeded };
  return [0, JSON.stringify(report, null, 2) + "\n", ""];
}

export function cmdList(dataDir: string): CommandResult {
  if (!isInitialized(dataDir)) {
    return [1, "", uninitializedError(dataDir)];
  }
  const documents = loadDocuments(dataDir).map((doc) => ({
    id: doc.id,
    title: doc.title,
    filename: doc.filename,
    lines: doc.lines.length,
  }));
  return [0, JSON.stringify({ documents }, null, 2) + "\n", ""];
}

export interface Citation {
  document: string;
  title: string;
  line: number;
  excerpt: string;
  score: number;
}

export function retrieve(documents: LoadedDocument[], question: string): Citation[] {
  const questionTokens = new Set(tokenize(question));
  const candidates: Citation[] = [];
  for (const doc of documents) {
    doc.lines.forEach((line, index) => {
      const stripped = line.trim();
      if (!stripped) {
        return;
      }
      const lineTokens = new Set(tokenize(stripped));
      let score = 0;
      for (const token of questionTokens) {
        if (lineTokens.has(token)) {
          score += 1;
        }
      }
      if (score > 0) {
        candidates.push({
          document: doc.id,
          title: doc.title,
          line: index + 1,
          excerpt: stripped,
          score,
        });
      }
    });
  }
  candidates.sort((a, b) => {
    if (a.score !== b.score) {
      return b.score - a.score;
    }
    if (a.document !== b.document) {
      return a.document < b.document ? -1 : 1;
    }
    return a.line - b.line;
  });
  return candidates.slice(0, MAX_CITATIONS);
}

// The model seam: a real assistant would generate prose here. This
// deterministic composer quotes the best citation instead, keeping the
// citation contract identical for both.
export function composeAnswer(citations: Citation[]): string {
  const [first, second] = citations;
  if (first === undefined) {
    return NO_MATCH_ANSWER;
  }
  let answer = `Based on "${first.title}" (line ${first.line}): ${first.excerpt}`;
  if (second !== undefined) {
    answer += ` See also "${second.title}" (line ${second.line}).`;
  }
  return answer;
}

export function cmdAsk(dataDir: string, question: string): CommandResult {
  if (!isInitialized(dataDir)) {
    return [1, "", uninitializedError(dataDir)];
  }
  const citations = retrieve(loadDocuments(dataDir), question);
  const report = { question, citations, answer: composeAnswer(citations) };
  return [0, JSON.stringify(report, null, 2) + "\n", ""];
}

// -------------------------------------------------------------------- serve

function healthPayload(dataDir: string): { status: string; documents: number } {
  return { status: "ok", documents: loadDocuments(dataDir).length };
}

function makeServer(dataDir: string): Server {
  return createServer((request, response) => {
    const url = new URL(request.url ?? "/", "http://127.0.0.1");
    const respond = (status: number, payload: unknown): void => {
      const body = JSON.stringify(payload, null, 2);
      response.writeHead(status, { "Content-Type": "application/json" });
      response.end(body);
    };
    process.stderr.write(`kb serve: "GET ${url.pathname}"\n`);
    if (url.pathname === "/health") {
      respond(200, healthPayload(dataDir));
    } else if (url.pathname === "/documents") {
      const [, out] = cmdList(dataDir);
      respond(200, JSON.parse(out));
    } else if (url.pathname === "/ask") {
      const [, out] = cmdAsk(dataDir, url.searchParams.get("q") ?? "");
      respond(200, JSON.parse(out));
    } else {
      respond(404, { error: "not found" });
    }
  });
}

async function cmdServe(dataDir: string, port: number, selfCheck: boolean): Promise<CommandResult> {
  if (!isInitialized(dataDir)) {
    return [1, "", uninitializedError(dataDir)];
  }
  const server = makeServer(dataDir);
  await new Promise<void>((resolve) => server.listen(port, "127.0.0.1", resolve));
  const address = server.address();
  const boundPort = typeof address === "object" && address !== null ? address.port : port;
  if (!selfCheck) {
    process.stderr.write(`kb serve: listening on http://127.0.0.1:${boundPort}\n`);
    await new Promise<void>((resolve) => server.on("close", resolve));
    return [0, "", ""];
  }
  try {
    const base = `http://127.0.0.1:${boundPort}`;
    const health = (await (await fetch(`${base}/health`)).json()) as Record<string, unknown>;
    const documents = (await (await fetch(`${base}/documents`)).json()) as {
      documents: unknown[];
    };
    const report = { self_check: { health, documents: documents.documents.length } };
    return [0, JSON.stringify(report, null, 2) + "\n", ""];
  } finally {
    server.close();
    server.closeAllConnections();
    server.unref();
  }
}

// ---------------------------------------------------------------------- cli

function parseFlags(argv: string[]): [Record<string, string>, string[]] {
  const flags: Record<string, string> = {};
  const positional: string[] = [];
  let index = 0;
  while (index < argv.length) {
    const arg = argv[index];
    if (arg === undefined) {
      break;
    }
    if (arg === "--data-dir" || arg === "--seed" || arg === "--port") {
      const value = argv[index + 1];
      if (value === undefined) {
        throw new Error(`missing value for ${arg}`);
      }
      flags[arg] = value;
      index += 2;
    } else if (arg === "--self-check") {
      flags[arg] = "true";
      index += 1;
    } else {
      positional.push(arg);
      index += 1;
    }
  }
  return [flags, positional];
}

export async function dispatch(argv: string[]): Promise<CommandResult> {
  const [command, ...rest] = argv;
  if (command === undefined) {
    return [2, "", USAGE];
  }
  let flags: Record<string, string>;
  let positional: string[];
  try {
    [flags, positional] = parseFlags(rest);
  } catch (error) {
    return [2, "", `error: ${(error as Error).message}`];
  }
  if (command === "init" && "--data-dir" in flags && positional.length === 0) {
    return cmdInit(flags["--data-dir"], flags["--seed"] ?? null);
  }
  if (command === "list" && "--data-dir" in flags && positional.length === 0) {
    return cmdList(flags["--data-dir"]);
  }
  const question = positional[0];
  if (command === "ask" && "--data-dir" in flags && positional.length === 1 && question !== undefined) {
    return cmdAsk(flags["--data-dir"], question);
  }
  if (command === "serve" && "--data-dir" in flags && positional.length === 0) {
    const portText = flags["--port"] ?? "0";
    if (!/^\d+$/.test(portText)) {
      return [2, "", `error: invalid port ${portText}`];
    }
    return cmdServe(flags["--data-dir"], Number(portText), "--self-check" in flags);
  }
  return [2, "", USAGE];
}

async function main(argv: readonly string[]): Promise<number> {
  const [exitCode, out, err] = await dispatch([...argv.slice(2)]);
  if (out) {
    process.stdout.write(out);
  }
  if (err) {
    process.stderr.write(err + "\n");
  }
  return exitCode;
}

if (import.meta.url === pathToFileURL(process.argv[1] ?? "").href) {
  main(process.argv).then((code) => {
    process.exitCode = code;
  });
}
